### PR TITLE
feat(useClickOutside): introduce

### PR DESCRIPTION
**Pull request by AugustinMauroy**
* [x] create useClickOutside, fails: on build
* [X] test useClickOutside (strange fails)
* [X] implement it 
## Related issue
#5348

### CODE REVIEW

File: hooks/useClickOutside.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { useEffect } from 'react';
+import type { RefObject, MouseEvent, TouchEvent } from 'react';
+
+// eslint-disable-next-line no-unused-vars
+type TypeHandler = (event: MouseEvent | TouchEvent) => void;
```
Comment (ovflowd) [inline hooks/useClickOutside.ts:5]: This should make ESLint happy

```suggestion
type TypeHandler = (_event: MouseEvent | TouchEvent) => void;
```

File: hooks/useClickOutside.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+export const useClickOutside = (
+  ref: RefObject<HTMLElement>,
+  handler: TypeHandler
+) => {
+  useEffect(() => {
+    const listener = (event: MouseEvent | TouchEvent) => {
+      if (!ref.current || ref.current.contains(event.target as Node)) {
```
Comment (ovflowd) [inline hooks/useClickOutside.ts:13]: Why a type casting needed? Also let's un-negate this condition:

```tsx
if (ref.current && !ref.current.contains(event.target)) {
  handler(event);
}
```

File: hooks/useClickOutside.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        return;
+      }
+      handler(event);
+    };
+
+    document.addEventListener('mousedown', listener);
```
Comment (ovflowd) [inline hooks/useClickOutside.ts]: Let's make these event listeners passive, just to make them non-blocking.

File: hooks/useClickOutside.ts [hunk 4]
```diff
@@ -2,7 +2,7 @@ import { useEffect } from 'react';
 import type { RefObject, MouseEvent, TouchEvent } from 'react';
 
 // eslint-disable-next-line no-unused-vars
```
Comment (ovflowd) [inline hooks/useClickOutside.ts:4]: You should be able to remove this eslint warning now.

File: hooks/useClickOutside.ts (hunk 1 above)
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts:5]: ESlint was not happy with this change 🤔
I'm searching deeply

File: hooks/useClickOutside.ts (hunk 2 above)
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts:13]: with your code we have:
"event.target is not attribuable to node"

File: hooks/useClickOutside.ts (hunk 3 above)
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: just that ?
```js
document.addEventListener('mousedown', listener, { passive: true });
document.addEventListener('touchstart', listener, { passive: true });
```
or
```js
document.addEventListener('mousedown', listener, true);
document.addEventListener('touchstart', listener, true});
```

File: hooks/useClickOutside.ts (hunk 1 above)
Comment (ovflowd) [reply hooks/useClickOutside.ts:5]: Probably Hinata's change wasn't merged yet to make eslint not complain about this.

File: hooks/useClickOutside.ts (hunk 2 above)
Comment (ovflowd) [reply hooks/useClickOutside.ts:13]: Try `event.currentTarget`

Comment (AugustinMauroy) [reply hooks/useClickOutside.ts:13]: Good i add this to my personal cheatsheet

Comment (AugustinMauroy) [reply hooks/useClickOutside.ts:13]: Oups say nothing when we use event.currentTarget we collapse dropdown when we click inside

Comment (AugustinMauroy) [reply hooks/useClickOutside.ts:13]: I search on web and other people use `as node`

File: hooks/useClickOutside.ts (hunk 4 above)
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts:4]: Question is refer [here](https://github.com/nodejs/nodejs.org/pull/5349#discussion_r1183592983)

**Comment by ovflowd** [comment]
**This PR got accidentally closed because the branch got deleted. Please feel free to reopen the PR.**

**Comment by ovflowd** [comment]
**This PR got accidentally closed because the branch got deleted. Please feel free to reopen the PR.**

**Comment by ovflowd** [comment]
I'm not sure you noticed, Augustin, but your branch is still completely broken.

As I advised before and will again, please read docs on how to properly rebase your branch or cherry pick your commits.

**Comment by AugustinMauroy** [comment]
i have follow this [docs](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/managing-commits/cherry-picking-a-commit)

**Comment by ovflowd** [comment]
Yes, but that's not all you need to do, let me give you some instructions.

1. You copy the commit hashes of the commits containing your actual changes
2. Checkout `main` (which should be safe)
3. Delete your local `major/website-redesign` branch as it is desynced (`git branch -D major/website-redesign`)
4. Check-out the `major/website-redesign` branch from upstream (this reduces the chances of getting desynced, if you use forks) `git switch -c upstream/major/website-redesign major/website-redesign`
5. Now create your PR branch again with same name `git checkout -b feat(useClickOutside)`
6. Now cherry-pick the commits with `git cherry-pick commit-1 commit-2 ...`
7. Force push to your branch on your fork `git push origin HEAD --force`

**Comment by AugustinMauroy** [comment]
Thanks for the advidce but im not using GitHub cil. I'm using GitHub. desktop

**Comment by ovflowd** [comment]
> Thanks for the advidce but im not using GitHub cil. I'm using GitHub. desktop

You can also use the regular Git CLI that comes with GitHub Desktop. These commands above are NOT GitHub CLI, are just plain Git

Also, again **you don't need to close this PR**, not sure why you're doing this 🤔

**Pull request closed without merging**